src/main/java/com/limechain/trie/structure/TrieStructure.java

```diff
@@ -1,30 +1,480 @@
+                    }
+
+                    parent = this.getNodeAtIndexInner(idx).parent;
+                }
+
+                // At the end, no nodes should be left.
+                return null;
+            }
+        );
+    }
+
+    /**
+     * @param nodeIndex index of a node
+     * @return the index of the next sibling of the node,
+     *         null if it's the last (lexicographically speaking) child if its parent, thus no next sibling exists.
+     */
+    private @Nullable Integer nextSibling(int nodeIndex) {
+        var parentIndexPair = this.getNodeAtIndexInner(nodeIndex).parent;
+
+        if (parentIndexPair == null) {
+            return null;
+        }
+
+        int parentIndex = parentIndexPair.parentNodeIndex();
+        Nibble childIndex = parentIndexPair.childIndexWithinParent();
+
+        var parent = this.getNodeAtIndexInner(parentIndex);
+
+        for (int idx = childIndex.asInt() + 1; idx < 16; ++idx) {
+            if (parent.childrenIndices[idx] != null) {
+                return parent.childrenIndices[idx];
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * @param key the full key as an iterator of nibbles
+     * @return the optional node handle for the given key, or empty if no node corresponds to the given key.
+     */
+    public Optional<NodeHandle<T>> existingNode(Nibbles key) {
+        return switch (this.existingNodeInner(key)) {
+            case ExistingNodeInnerResult.Found found -> Optional.of(
+                    NodeHandle
+                        .<T>getConstructor(found.hasStorageValue())
+                        .apply(this, found.nodeIndex())
+            );
+            case ExistingNodeInnerResult.NotFound ignored -> Optional.empty();
+        };
+    }
+
+    private ExistingNodeInnerResult existingNodeInner(final Nibbles key) {
+        if (this.rootIndex == null) {
+            return new ExistingNodeInnerResult.NotFound(null);
+        }
+
+        int currentIndex = this.rootIndex;
+
+        assert this.getNodeAtIndexInner(currentIndex).parent == null : "Root's parent index must be null.";
+
+        ExistingNodeInnerResult.NotFound.ClosestAncestor closestAncestor = null;
+
+        Iterator<Nibble> keyIter = key.iterator();
+        while (true) {
+            TrieNode<T> current = this.nodes.get(currentIndex);
+
+            // First, we must remove `current`'s partial key from `key`, making sure that they
+            // match.
+            for (Nibble nibble : current.partialKey) {
+                if (!keyIter.hasNext() || !keyIter.next().equals(nibble)) {
+                    return new ExistingNodeInnerResult.NotFound(closestAncestor);
+                }
+            }
+
+            // At this point, the tree traversal cursor (the already consumed part of the `key` iterator)
+            // exactly matches `current`.
+
+            // TODO: clone the `key` iterator more sanely... we're currently copying the whole underlying data source
```

Comment:
To my surprise, I haven't found a more elegant solution to this problem. Really, what we need is simply to copy the iterator itself, but Java doesn't seem to provide an easy way to do it as it's not the intended use of iterators... so I'm just copying the remaining nibbles everytime anew, just to start iterating them from where we've left off.

P.S. Originally the constructor of `NotFound.ClosestAncestor` accepted the iterator only, since that suffices, but since we were copying anyway, I made it own the entire underlying `Nibbles` collection.

Reply:
I don't know if there is a more "sane" approach for copying iterators. You can do a quick search about it but if nothing comes up just delete the todo.

Reply:
I've already searched for it, the simple answer is "iterators are not meant to be copied", except that's exactly what I want. So the most commonly proposed solution is namely that: collect the remainder of the iterator into a list and continue iterating from the new list 😄 

Are we okay with transforming the TODO to a NOTE, since that's a possible future refactor point I think is worth considering...?